rs/messaging/src/state_machine.rs

```diff
@@ -183,9 +188,16 @@ impl StateMachine for StateMachineImpl {
 
         let since = Instant::now();
         // Postprocess the state and consolidate the Streams.
-        let state_after_stream_builder = self.stream_builder.build_streams(state_after_execution);
+        let mut state_after_stream_builder =
+            self.stream_builder.build_streams(state_after_execution);
         self.observe_phase_duration(PHASE_MESSAGE_ROUTING, &since);
 
+        let since = Instant::now();
+        // Shed some messages if above the best-effort message memory limit.
+        state_after_stream_builder
+            .enforce_best_effort_message_limit(self.best_effort_message_memory_capacity);
```

Comment:
Was it an explicit decision to shed after building the stream? I wonder whether it might make sense to shed before?

Reply:
I didn't think of it in those specific terms -- before routing vs after routing. I did it at the very end of the round, so in-between rounds we're always in a "valid" state.

Streams are limited in size (10 MB per stream; although once we have more subnets we may also want to impose a global limit), so I don't think that shedding before vs. after is going to make a huge difference (we're talking shedding messages above something like 5 GB; from the couple of canisters with the highest usage; while we're likely going to route at most a few dozen MB).

Are you thinking of a specific reason why it might make more sense to do this before routing?

Reply:
No, I was just curious whether there might be a reason I'm missing. Thanks!